### ISSUE TITLE
Crypto.Cipher.PKCS1_v1_5 is most likely vulnerable to timing variant of Bleichenbacher attack

### ISSUE THREAD

**Issue by tomato42** [Legrandin/pycryptodome#481]
(I haven't found a security policy suggesting submission of security bugs should happen in a specific way so I'm posting a regular issue)

The code in PKCS#1 v1.5 is most likely vulnerable to the timing variant of the Bleichenbacher attack:
https://github.com/Legrandin/pycryptodome/blob/67c3f12990ec86d288b6ff5e6b2d1a43a6a1a099/lib/Crypto/Cipher/PKCS1_v1_5.py#L159-L177

Issues I can identify:
 1. The RSA decryption code uses non side-channel free multiply to undo blinding: https://github.com/Legrandin/pycryptodome/blob/67c3f12990ec86d288b6ff5e6b2d1a43a6a1a099/lib/Crypto/PublicKey/RSA.py#L170
 2. the `long_to_bytes` depends on bit length of the decrypted message: https://github.com/Legrandin/pycryptodome/blob/67c3f12990ec86d288b6ff5e6b2d1a43a6a1a099/lib/Crypto/Util/number.py#L401-L403
 3. the search for the null separator byte is not guaranteed to inspect all bytes of the decrypted message (`find()` will return as soon as it finds the searched-for byte): https://github.com/Legrandin/pycryptodome/blob/67c3f12990ec86d288b6ff5e6b2d1a43a6a1a099/lib/Crypto/Cipher/PKCS1_v1_5.py#L173
 4. the selection of the returned value uses a jump: https://github.com/Legrandin/pycryptodome/blob/67c3f12990ec86d288b6ff5e6b2d1a43a6a1a099/lib/Crypto/Cipher/PKCS1_v1_5.py#L174-L177
5. in case the real value is returned, it's subscripted, creating a new object in the process: https://github.com/Legrandin/pycryptodome/blob/67c3f12990ec86d288b6ff5e6b2d1a43a6a1a099/lib/Crypto/Cipher/PKCS1_v1_5.py#L177
6. the `if` condition uses less-than, `bytes.startswith()`, and short circuit (`or`) in branch selection, all of which are data-dependant: https://github.com/Legrandin/pycryptodome/blob/67c3f12990ec86d288b6ff5e6b2d1a43a6a1a099/lib/Crypto/Cipher/PKCS1_v1_5.py#L174
7. the calling code will access different parts of memory in case padding check failed (may not be vulnerable to simple timing, but most likely will be vulnerable to FLUSH+RELOAD and similar attacks)

While some of those issues can be "fixed" (by using bit operations on integers for the `if`, not bool's), they will still leave a timing side-channel when implemented using native python `int()`'s: https://securitypitfalls.wordpress.com/2018/08/03/constant-time-compare-in-python/

**Comment by Legrandin** [comment]
As a note, this is explicitly called out in the documentation:

https://www.pycryptodome.org/en/latest/src/cipher/pkcs1_v1_5.html

RSAES-PKCS1-v1_5 is inherently vulnerable to side-channel attacks, and even fixing it in C or assembly isn't easy. People should use OAEP instead. The current implementation is really there only for people that can use in an environment where it can't be possibly used an oracle and hopefully only to decrypt legacy data.

For future reference, you can drop sensitive disclosures to security@pycryptodome.org  .

**Comment by tomato42** [comment]
> As a note, this is explicitly called out in the documentation:
> 
> https://www.pycryptodome.org/en/latest/src/cipher/pkcs1_v1_5.html

my reading of the documentation is more along the lines of "use this API carefully, as improper use of it will make your code vulnerable" than "this API is insecure, no matter how you use it and whichever workarounds you implement"

while in reality it's the latter 

> RSAES-PKCS1-v1_5 is inherently vulnerable to side-channel attacks, and even fixing it in C or assembly isn't easy. People should use OAEP instead. The current implementation is really there only for people that can use in an environment where it can't be possibly used an oracle and hopefully only to decrypt legacy data.

doesn't OAEP use the exact same problematic code as the one singled out in problems 1 and 2?
https://github.com/Legrandin/pycryptodome/blob/67c3f12990ec86d288b6ff5e6b2d1a43a6a1a099/lib/Crypto/Cipher/PKCS1_OAEP.py#L171-L174
 
> For future reference, you can drop sensitive disclosures to [security@pycryptodome.org](mailto:security@pycryptodome.org) .

could you create a security policy file or a security section in documentation with that info?

**Comment by tomato42** [comment]
It looks like the `sentinel` was introduced in 2011, in https://github.com/Legrandin/pycryptodome/commit/02103e2a5aca97b299b63723fb6752c2cbc00b23 with mostly identical code as it is now, so basically all versions are vulnerable.

**Comment by tomato42** [comment]
@Legrandin this code is the same in pycrypto, so I'll be asking for a CVE for it, should I ask for one for pycryptodome too?

**Comment by Legrandin** [comment]
> my reading of the documentation is more along the lines of "use this API carefully, as improper use of it will make your code
> vulnerable" than "this API is insecure, no matter how you use it and whichever workarounds you implement"
> while in reality it's the latter

Point taken. We can make that more explicit.

> doesn't OAEP use the exact same problematic code as the one singled out in problems 1 and 2?

I think you need to make a distinction between leakage of key material and leakage of plaintext, because the threat models are different.

The former is much more severe. Leakage of plaintext is recognized a lesser concern. More importantly, unlike the former, plaintext leakage is not only limited to the crypto library alone. The rest of the application (caller) that processes the returned plaintext value needs to be made SCA-resistant too. In other words, while I agree that it's probably simple enough to fix the lines you point at, I think that a) leakage of PT at those points is very limited b) most probably, most of the leakage is happening down the line b) you should not be using Python in the first place if that is your threat model...

As a comment, I even think that the title of this issue is wrong, as it seems to suggest that RSAES-PKCS1-v1_5 can be made leak resistant. I am not confident it can, no matter how you implement it. I'd rather see people give up on it, and move to OAEP or (better) ECC.

> could you create a security policy file or a security section in documentation with that info?

Yes, that will happen soon.

**Comment by tomato42** [comment]
> > doesn't OAEP use the exact same problematic code as the one singled out in problems 1 and 2?
> 
> I think you need to make a distinction between leakage of key material and leakage of plaintext, because the threat models are different.

yes, and leakage of RSA key is prevented by use of bliding, that's happening and it's done correctly, but leakage of plaintext is still a valid threat vector. And Bleichenbacher attack never was used to extract the key, only to perform private key operations...

> The former is much more severe. Leakage of plaintext is recognized a lesser concern. More importantly, unlike the former, plaintext leakage is not only limited to the crypto library alone. The rest of the application (caller) that processes the returned plaintext value needs to be made SCA-resistant too. In other words, while I agree that it's probably simple enough to fix the lines you point at, I think that a) leakage of PT at those points is very limited b) most probably, most of the leakage is happening down the line b) you should not be using Python in the first place if that is your threat model...

ok, but is pycryptodome supposed to be secure against side-channel attacks or not? Because if that's not the threat model for pycryptodome then I don't see it documented. If it is supposed to be secure against side-channel attacks, then OAEP is still most likely vulnerable (yes, it may be hard to exploit, but that's a different thing).

yes, the caller needs to be careful not to reveal secrets, but if it does that and the library it uses still leaks, then it's still vulnerable, and there's nothing it can do to fix it (building on sand and all that)
 
> As a comment, I even think that the title of this issue is wrong, as it seems to suggest that RSAES-PKCS1-v1_5 can be made leak resistant. I am not confident it can, no matter how you implement it. I'd rather see people give up on it, and move to OAEP or (better) ECC.

in pure python? no, I don't think you can make it side-channel secure. But in general, yes, I think it's possible to make it secure, albeit it's very non-trivial (just like side-channel secure ECC or fast side channel secure AES in software is hard to do)

finally, I do agree that people should use OAEP instead, but that's something we've been telling people for 20 years, and they still continue to use PKCS#1 v1.5, so clearly, just telling people to stop using it is not working

**Comment by eslerm** [comment]
> I think you need to make a distinction between leakage of key material and leakage of plaintext, because the threat models are different.

I can confirm that PyCryptodome's PKCS#1 v1.5 decrypt is a behavioral [Bleichenbacher](https://www.iacr.org/cryptodb/data/paper.php?pubkey=1037) oracle and have used it to recover the plaintext of a 4096-bit RSA ciphertext. The same oracle can be used to forge signatures.
